Add  basic PEFT support to train script + record module

githubnemo opened the pull request:
Hello from PEFT :wave:

## What this does

This change introduces basic support for PEFT adapter methods (i.e. everything that has a `rank` parameter, LoRA, BONE, LoHa, LoKr, etc). In the future it probably makes sense to add a more generic `--peft.adapter_config` parameter that allows for passing a PEFT `adapter_config.json` or possibly its contents as a string to support methods that need a more complex setup.

Another benefit of having PEFT is that remixing models is much easier since you only download/upload small adapters (~10MB) instead of whole models (~1GB) which is good for iteration, I suppose.

## General considerations

It is currently only possible to train pre-trained policies with PEFT. Training policies from scratch and configuring PEFT will raise an error. Arguably, fine-tuning pre-trained policies is the most common use-case. There are probably some cases where it makes sense to be able to train a policy from scratch but I'd rather implement these with a concrete example in mind. Also being able to train from scratch is likely to yield bad results and is a foot-gun that is easily employed.

## How it was tested

These changes were tested on a small custom dataset for the SO101. There it was tested that SmolVLA + PEFT roughly performs equally to SmolVLA for the same training time budget. I also made sure that the RTC example works with PEFT.

LIBERO was used for more systematic testing of the PEFT integration, although mostly on SmolVLA. Here are the results:

Category | Method | Rank | BS | LR | Steps | Sched. Decay | Avg Reward
-- | -- | -- | -- | -- | -- | -- | --
BASELINE | Baseline (Pretrained) | - | 32 | $1e-4$ | 100k | $2.5e-6$ | 0.79
PEFT | LoRA + FFT | 128 | 32 | $1e-2$ | 100k | $2.5e-6$ | 0.65
PEFT | LoRA + FFT | 64 | 32 | $1e-2$ | 100k | $1e-3$ | 0.66
PEFT | LoRA + FFT | 64 | 32 | $1e-3$ | 100k | $2.5e-6$ | 0.68
PEFT | LoRA (Only Adapter) | 64 | 32 | $1e-3$ | 100k | $1e-3$ | 0.68
PEFT | LoRA (All Adapter) | 64 | 32 | $1e-3$ | 100k | $1e-3$ | 0.45
PEFT | LoRA (All Adapter) | 64 | 32 | $1e-3$ | 100k | $1e-4$ | 0.73

Notes:
- All LoRA models base on `lerobot/smolvla_base`
- $2.5e-6$ is the default lr scheduler decay target of SmolVLA
- LoRA + FFT means that q/v projections of `lm_expert` are adapted but state and
action projections are tuned fully
- LoRA (Only Adapter) means that just the q/v projections are fine-tuned
- LoRA (All Adapter) means that q/v projections and state/action projections are
adapted on LoRA

The "LoRA (All Adapter)" configuration is what is currently implemented as default since it yielded the best results and is also the most parameter efficient.

Training script baseline:

```bash
lerobot-train \
 --policy.path=lerobot/smolvla_base \
 --dataset.repo_id=HuggingFaceVLA/libero \
 --policy.output_features=null \
 --policy.input_features=null \
 --env.type=libero \
 --env.task=libero_spatial \
 --output_dir=./outputs/"${repo_name}" \
 --policy.optimizer_lr=1e-4 \
 --steps=100000 \
 --batch_size=32
```

The difference of PEFT and baseline training invocations is:
```bash
 --peft.method_type=LORA
 --peft.r=$rank
```

I've also tested pi05 insofar that I wanted to compare the max. GPU memory consumption for fine-tuning on baseline and with PEFT to see if PEFT can lower the VRAM usage. PEFT does that. On a NVIDIA L40S (48GB VRAM) I can fine-tune pi05 with PEFT and batch size 4 while I run out-of-memory when trying to do the same (even with batch size 1) without PEFT.

## How to checkout & try?

Currently `lerobot-train`, `lerobot-eval` and `lerobot-record` can deal with policies trained with PEFT. I've also adapted the RTC example to detect a PEFT model. Everything that uses `make_policy` should be able to support it out of the box but many examples probably don't yet.

Commit:
Add basic support for PEFT adapter methods

This changes adds support for training policies with much less parameters
by applying adapter methods such as LoRA on specific parts of the policies
and therefore possibly higher learning rates / batch sizes.

To make this as accessible as possible I thought it useful to provide
defaults for `target_modules` and `modules_to_save`. Currently only SmolVLA
has such defaults but when we agree that this change is useful I will set
out to generate more such defaults. While the user can override these
settings, they are expected to only change the peft_method, rank and init_type
parameters.

Commit:
Implement loading of PEFT adapters

Loading a PEFT adapter is currently done by initializing a policy with default config
and then applying the adapter on the resulting model. This has the obvious drawback
that any configurations done during training are not applied in the adapted model.

Currently the `use_peft` attribute of `PreTrainedConfig` is only set during loading
to signal the following code that it has to deal with a PEFT adapter. However
we could imagine a scenario where this is already set at training time and stored
alongside the adapter.

Commit:
Store policy config alongside PEFT checkpoint

Before this change the PEFT-wrapped policy did not save the policy's config
alongside the adapter config / weights which prevented us from changing the
policy config. Now the policy config is saved both in full training and PEFT
training.

This change makes loading the PEFT policy adapter much easier as well.

Commit:
Merge branch 'main' of github.com:huggingface/lerobot into feature/basic-peft-support

Commit:
Add default config for ACT

Commit:
Support targets like `all-linear`

Commit:
Formatting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
Merge branch 'main' of github.com:huggingface/lerobot into feature/basic-peft-support

Commit:
Fix failing tests

Commit:
Merge remote-tracking branch 'hf/main' into feature/basic-peft-support

pkooij reviewed (commented):
Looks good, I am now running training, just left a few small comments!

Commit:
Remove PEFT compatibility changes in config

We'll wait for the PEFT release that fixes this for good.

Commit:
Remove `use_peft` parameter from training script

Instead we make the PEFT config optional which has the same effect.

Commit:
Log adapter config to WandB

Commit:
Better documentation for CLI arguments

Commit:
Don't unload & merge the PEFT model

This can make things hard when using quantized layers (user expects quantized base layers with
unquantized adapters for example, merging defaults to upcast the layers leading to higher
memory).

Commit:
Correct way of identifying when to save config

Commit:
Add CLI end-to-end tests

Currently there don't seem to be any way to test the CLI commands.
Since this change mostly happens in those I thought it best to add
a way to test these commands end-to-end.

More integrated commands like `lerobot-record` need patching but
standalone commands like training seem to work fine.

Commit:
Update default targets

Removed ACT since it doesn't make sense to fine-tune ACT without having it pretrained beforehand.
SmolVLA and Pi0/0.5 are much more senseful targets.

BenjaminBossan commented:
> it looks like peft is trying to use the config object as a dict, but it is not a dict. is there a specific version of peft that needs to be installed?

@githubnemo Does this require the latest PEFT installed from source?

Commit:
Merge remote-tracking branch 'origin/main' into feature/basic-peft-support

Commit:
Merge remote-tracking branch 'hf/main' into feature/basic-peft-support

Commit:
Clean up loading code

- Centralized instantiation of the PEFT wrapper in `make_policy` for inference
  (e.g. in `lerobot-record`)
- Training a PEFT policy also sets `cfg.use_peft` so that all inference code loading
  the policy can rely on that attribute to identify if PEFT loading is needed
- Modified RTC example to also include PEFT policies. Mostly because this is an example
  I'm currently exploring.

Commit:
Make sure push_to_hub works

Since PEFT only wraps `push_to_hub` and not `push_model_to_hub`, the reference
to `self` in `policy.push_model_to_hub` is the unwrapped policy which, of course,
doesn't know anything about PEFT.

To make the upload process aware of PEFT, we pass the unwrapped policy down to
`push_model_to_hub` as a kwarg. This is not ideal but I think it is the best way
for now.

Commit:
formatting

Commit:
Warn when encountering from-scratch-training

Commit:
Merge remote-tracking branch 'hf/main' into feature/basic-peft-support

Commit:
Revamp pretrained model loading

There were quite a few factors that convinced me that the status quo
is able to load pretrained models from the PEFT adapter config but
in fact that didn't work.

This commit fixes the following things:
- policies wrapped in PEFT will now have a `name_or_path` attribute
  containing the name or path of the pretrained model we're fine-tuning
- we further assume that SmolVLA without `pretrained_path` and
  `load_vlm_weights==False` must be an user-side error
- we assume that using PEFT on from-scratch-policies must be
  an user-side-error

Commit:
Make it possible to unset policy features

This is necessary to train pre-trained policies on new datasets so that the
features are inferred from the new dataset and not from the pretrained
policy.

Commit:
Use correct loading for PEFT in RTC example

Commit:
Make it possible to use PeftModels in eval

Commit:
Add test checking that PEFT actually reduces params

Commit:
Adapt state/action projections instead of full-finetuning

There doesn't seem to be a benefit to fully fine-tune these layers
over just adapting them, so we do that instead.

Commit:
Merge remote-tracking branch 'hf/main' into feature/basic-peft-support

Commit:
Disallow PEFT training on non-pretrained policies

At first I thought it would make sense to have this feature
in case you want to fine-tune a pre-trained section but in the
end it makes more trouble than it's worth.

It's still possible to allow this in the future when a concrete
need arises.

Commit:
Add basic documentation

Commit:
Merge remote-tracking branch 'hf/main' into feature/basic-peft-support

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/lerobot/pr_1411). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

githubnemo commented:
I've updated the description. A lot of stuff happened in the meantime. I think this is ready for review now. @pkooij

Commit:
Formatting

Commit:
Add peft as extra dependency, mark tests

Fast tests currently fail because of the missing dependency.

pkooij committed:
Merge branch 'main' into feature/basic-peft-support

pkooij committed:
Merge branch 'main' into feature/basic-peft-support

Commit:
Fix pre-commit issues

Commit:
Add walx <> peft conflict for uv

Commit:
Exclude peft from pi install for now

pkooij committed:
Merge branch 'main' into feature/basic-peft-support

cjd-hub commented:
Hello! Thank you for your work. The importance of LoRA for scenarios with insufficient GPU resources cannot be overstated. I would like to ask: when performing LoRA fine‑tuning based on pi0 or pi0.5, is it sufficient to simply install lerobot[peft] and add parameters like --peft.method_type=LORA --peft.r=64 (assuming the code modifications have already been synchronized with this post)?

githubnemo commented:
> I would like to ask: when performing LoRA fine‑tuning based on pi0 or pi0.5, is it sufficient to simply install lerobot[peft] and add parameters like --peft.method_type=LORA --peft.r=64 (assuming the code modifications have already been synchronized with this post)?

Yes, that should be sufficient. Normally you expect to define layers that are targeted by the adapters - this is still the case but there are model-specific defaults. For pi05 they are defined on the policy in `_get_default_peft_targets`:

```
        common_projections = (
            "state_proj|action_in_proj|action_out_proj|action_time_mlp_in|action_time_mlp_out"
        )
        target_modules = rf"(.*\.gemma_expert\..*\.self_attn\.(q|v)_proj|model\.({common_projections}))"
```

Some more details may be found in the [docs](https://huggingface.co/docs/lerobot/en/peft_training).
